README.md
![MathPHP Logo](https://github.com/markrogoyski/math-php/blob/master/docs/image/MathPHPLogo.png?raw=true)

### MathPHP - Powerful Modern Math Library for PHP

The only library you need to integrate mathematical functions into your applications. It is a self-contained library in
pure PHP with no external dependencies.

[![Coverage Status](https://coveralls.io/repos/github/markrogoyski/math-php/badge.svg?branch=master)](https://coveralls.io/github/markrogoyski/math-php?branch=master)
[![License](https://poser.pugx.org/markrogoyski/math-php/license)](https://packagist.org/packages/markrogoyski/math-php)

Features
--------

* [Algebra](#algebra)
* [Arithmetic](#arithmetic)
* Expression
    - [Polynomial](#expression---polynomial)
* [Finance](#finance)
* Functions
    - [Map](#functions---map---single-array)
    - [Special Functions](#functions---special-functions)
* Information Theory
    - [Entropy](#information-theory---entropy)
* Linear Algebra
    - [Matrix](#linear-algebra---matrix)
    - [Vector](#linear-algebra---vector)
* Numbers
    - [Arbitrary Integer](#number---arbitrary-length-integers)
    - [Complex](#number---complex-numbers)
    - [Quaternion](#number---quaternion)
    - [Rational](#number---rational-numbers)
* Number Theory
    - [Integers](#number-theory---integers)
* Numerical Analysis
    - [Interpolation](#numerical-analysis---interpolation)
    - [Numerical Differentiation](#numerical-analysis---numerical-differentiation)
    - [Numerical Integration](#numerical-analysis---numerical-integration)
    - [Root Finding](#numerical-analysis---root-finding)
* Probability
    - [Combinatorics](#probability---combinatorics)
    - Distributions
        * [Continuous](#probability---continuous-distributions)
        * [Discrete](#probability---discrete-distributions)
        * [Multivariate](#probability---multivariate-distributions)
        * [Tables](#probability---distribution-tables)
* [Sample Data](#sample-data)
* [Search](#search)
* Sequences
    - [Basic](#sequences---basic)
    - [Advanced](#sequences---advanced)
    - [NonInteger](#sequences---non-integer)
* [Set Theory](#set-theory)
* Statistics
    - [ANOVA](#statistics---anova)
    - [Averages](#statistics---averages)
    - [Circular](#statistics---circular)
    - [Correlation](#statistics---correlation)
    - [Descriptive](#statistics---descriptive)
    - [Distance](#statistics---distance)
    - [Distributions](#statistics---distributions)
    - [Divergence](#statistics---divergence)
    - [Effect Size](#statistics---effect-size)
    - [Experiments](#statistics---experiments)
    - [Kernel Density Estimation](#statistics---kernel-density-estimation)
    - Multivariate
        * [PCA (Principal Component Analysis)](#statistics---multivariate---principal-component-analysis)
        * [PLS (Partial Least Squares Regression)](#statistics---multivariate---partial-least-squares-regression)
    - [Outlier](#statistics---outlier)
    - [Random Variables](#statistics---random-variables)
    - [Regressions](#statistics---regressions)
    - [Significance Testing](#statistics---significance-testing)
* [Trigonometry](#trigonometry)

Setup
-----

Add the library to your `composer.json` file in your project:

```javascript
{
  "require": {
      "markrogoyski/math-php": "2.*"
  }
}
```

Use [composer](http://getcomposer.org) to install the library:

```bash
$ php composer.phar install
```

Composer will install MathPHP inside your vendor folder. Then you can add the following to your
.php files to use the library with Autoloading.

```php
require_once __DIR__ . '/vendor/autoload.php';
```

Alternatively, use composer on the command line to require and install MathPHP:

```
$ php composer.phar require markrogoyski/math-php:2.*
```

### Minimum Requirements

* PHP 7.2

Note: For PHP 7.0 and 7.1, use v1.0 (`markrogoyski/math-php:1.*`)

Usage
-----

### Algebra

```php
use MathPHP\Algebra;

// Greatest common divisor (GCD)
$gcd = Algebra::gcd(8, 12);

// Extended greatest common divisor - gcd(a, b) = a*a' + b*b'
$gcd = Algebra::extendedGcd(12, 8); // returns array [gcd, a', b']

// Least common multiple (LCM)
$lcm = Algebra::lcm(5, 2);

// Factors of an integer
$factors = Algebra::factors(12); // returns [1, 2, 3, 4, 6, 12]

// Linear equation of one variable: ax + b = 0
[$a, $b] = [2, 4]; // 2x + 4 = 0
$x       = Algebra::linear($a, $b);

// Quadratic equation: ax² + bx + c = 0
[$a, $b, $c] = [1, 2, -8]; // x² + 2x - 8
[$x₁, $x₂]   = Algebra::quadratic($a, $b, $c);

// Discriminant: Δ = b² - 4ac
[$a, $b, $c] = [2, 3, 4]; // 3² - 4(2)(4)
$Δ           = Algebra::discriminant($a, $b, $c);

// Cubic equation: z³ + a₂z² + a₁z + a₀ = 0
[$a₃, $a₂, $a₁, $a₀] = [2, 9, 3, -4]; // 2x³ + 9x² + 3x -4
[$x₁, $x₂, $x₃]      = Algebra::cubic($a₃, $a₂, $a₁, $a₀);

// Quartic equation: a₄z⁴ + a₃z³ + a₂z² + a₁z + a₀ = 0
[$a₄, $a₃, $a₂, $a₁, $a₀] = [1, -10, 35, -50, 24]; // z⁴ - 10z³ + 35z² - 50z + 24 = 0
[$z₁, $z₂, $z₃, $z₄]      = Algebra::quartic($a₄, $a₃, $a₂, $a₁, $a₀);
```

### Arithmetic

```php
use MathPHP\Arithmetic;

$√x  = Arithmetic::isqrt(8);     // 2 Integer square root
$³√x = Arithmetic::cubeRoot(-8); // -2
$ⁿ√x = Arithmetic::root(81, 4);  // nᵗʰ root (4ᵗʰ): 3

// Sum of digits
$digit_sum    = Arithmetic::digitSum(99);    // 18
$digital_root = Arithmetic::digitalRoot(99); // 9

// Equality of numbers within a tolerance
$x = 0.00000003458;
$y = 0.00000003455;
$ε = 0.0000000001;
$almostEqual = Arithmetic::almostEqual($x, $y, $ε); // true

// Copy sign
$magnitude = 5;
$sign      = -3;
$signed_magnitude = Arithmetic::copySign($magnitude, $sign); // -5

// Modulo (Differs from PHP remainder (%) operator for negative numbers)
$dividend = 12;
$divisor  = 5;
$modulo   = Arithmetic::modulo($dividend, $divisor);  // 2
$modulo   = Arithmetic::modulo(-$dividend, $divisor); // 3
```

### Expression - Polynomial

```php
use MathPHP\Expression\Polynomial;

// Polynomial x² + 2x + 3
$coefficients = [1, 2, 3]
$polynomial   = new Polynomial($coefficients);

// Evaluate for x = 3
$x = 3;
$y = $polynomial($x);  // 18: 3² + 2*3 + 3

// Calculus
$derivative = $polynomial->differentiate();  // Polynomial 2x + 2
$integral   = $polynomial->integrate();      // Polynomial ⅓x³ + x² + 3x

// Arithmetic
$sum        = $polynomial->add($polynomial);       // Polynomial 2x² + 4x + 6
$sum        = $polynomial->add(2);                 // Polynomial x² + 2x + 5
$difference = $polynomial->subtract($polynomial);  // Polynomial 0
$difference = $polynomial->subtract(2);            // Polynomial x² + 2x + 1
$product    = $polynomial->multiply($polynomial);  // Polynomial x⁴ + 4x³ + 10x² + 12x + 9
$product    = $polynomial->multiply(2);            // Polynomial 2x² + 4x + 6
$negated    = $polynomial->negate();               // Polynomial -x² - 2x - 3

// Data
$degree       = $polynomial->getDegree();        // 2
$coefficients = $polynomial->getCoefficients();  // [1, 2, 3]

// String representation
print($polynomial);  // x² + 2x + 3

// Roots
$polynomial = new Polynomial([1, -3, -4]);
$roots      = $polynomial->roots();         // [-1, 4]

// Companion matrix
$companion = $polynomial->companionMatrix();
```

### Finance

```php
use MathPHP\Finance;

// Financial payment for a loan or annuity with compound interest
$rate          = 0.035 / 12; // 3.5% interest paid at the end of every month
$periods       = 30 * 12;    // 30-year mortgage
$present_value = 265000;     // Mortgage note of $265,000.00
$future_value  = 0;
$beginning     = false;      // Adjust the payment to the beginning or end of the period
$pmt           = Finance::pmt($rate, $periods, $present_value, $future_value, $beginning);

// Interest on a financial payment for a loan or annuity with compound interest.
$period = 1; // First payment period
$ipmt   = Finance::ipmt($rate, $period, $periods, $present_value, $future_value, $beginning);

// Principle on a financial payment for a loan or annuity with compound interest
$ppmt = Finance::ppmt($rate, $period, $periods, $present_value, $future_value = 0, $beginning);

// Number of payment periods of an annuity.
$periods = Finance::periods($rate, $payment, $present_value, $future_value, $beginning);

// Annual Equivalent Rate (AER) of an annual percentage rate (APR)
$nominal = 0.035; // APR 3.5% interest
$periods = 12;    // Compounded monthly
$aer     = Finance::aer($nominal, $periods);

// Annual nominal rate of an annual effective rate (AER)
$nomial = Finance::nominal($aer, $periods);

// Future value for a loan or annuity with compound interest
$payment = 1189.97;
$fv      = Finance::fv($rate, $periods, $payment, $present_value, $beginning)

// Present value for a loan or annuity with compound interest
$pv = Finance::pv($rate, $periods, $payment, $future_value, $beginning)

// Net present value of cash flows
$values = [-1000, 100, 200, 300, 400];
$npv    = Finance::npv($rate, $values);

// Interest rate per period of an annuity
$beginning = false; // Adjust the payment to the beginning or end of the period
$rate      = Finance::rate($periods, $payment, $present_value, $future_value, $beginning);

// Internal rate of return
$values = [-100, 50, 40, 30];
$irr    = Finance::irr($values); // Rate of return of an initial investment of $100 with returns of $50, $40, and $30

// Modified internal rate of return
$finance_rate      = 0.05; // 5% financing
$reinvestment_rate = 0.10; // reinvested at 10%
$mirr              = Finance::mirr($values, $finance_rate); // rate of return of an initial investment of $100 at 5% financing with returns of $50, $40, and $30 reinvested at 10%

// Discounted payback of an investment
$values  = [-1000, 100, 200, 300, 400, 500];
$rate    = 0.1;
$payback = Finance::payback($values, $rate); // The payback period of an investment with a $1,000 investment and future returns of $100, $200, $300, $400, $500 and a discount rate of 0.10

// Profitability index
$values              = [-100, 50, 50, 50];
$profitability_index = Finance::profitabilityIndex($values, $rate); // The profitability index of an initial $100 investment with future returns of $50, $50, $50 with a 10% discount rate
```

### Functions - Map - Single Array

```php
use MathPHP\Functions\Map;

$x = [1, 2, 3, 4];

$sums        = Map\Single::add($x, 2);      // [3, 4, 5, 6]
$differences = Map\Single::subtract($x, 1); // [0, 1, 2, 3]
$products    = Map\Single::multiply($x, 5); // [5, 10, 15, 20]
$quotients   = Map\Single::divide($x, 2);   // [0.5, 1, 1.5, 2]
$x²          = Map\Single::square($x);      // [1, 4, 9, 16]
$x³          = Map\Single::cube($x);        // [1, 8, 27, 64]
$x⁴          = Map\Single::pow($x, 4);      // [1, 16, 81, 256]
$√x          = Map\Single::sqrt($x);        // [1, 1.414, 1.732, 2]
$∣x∣         = Map\Single::abs($x);         // [1, 2, 3, 4]
$maxes       = Map\Single::max($x, 3);      // [3, 3, 3, 4]
$mins        = Map\Single::min($x, 3);      // [1, 2, 3, 3]
$reciprocals = Map\Single::reciprocal($x);  // [1, 1/2, 1/3, 1/4]
```

### Functions - Map - Multiple Arrays

```php
use MathPHP\Functions\Map;

$x = [10, 10, 10, 10];
$y = [1,   2,  5, 10];

// Map function against elements of two or more arrays, item by item (by item ...)
$sums        = Map\Multi::add($x, $y);      // [11, 12, 15, 20]
$differences = Map\Multi::subtract($x, $y); // [9, 8, 5, 0]
$products    = Map\Multi::multiply($x, $y); // [10, 20, 50, 100]
$quotients   = Map\Multi::divide($x, $y);   // [10, 5, 2, 1]
$maxes       = Map\Multi::max($x, $y);      // [10, 10, 10, 10]
$mins        = Map\Multi::mins($x, $y);     // [1, 2, 5, 10]

// All functions work on multiple arrays; not limited to just two
$x    = [10, 10, 10, 10];
$y    = [1,   2,  5, 10];
$z    = [4,   5,  6,  7];
$sums = Map\Multi::add($x, $y, $z); // [15, 17, 21, 27]
```

### Functions - Special Functions

```php
use MathPHP\Functions\Special;

// Gamma function Γ(z)
$z = 4;
$Γ = Special::gamma($z);
$Γ = Special::gammaLanczos($z);   // Lanczos approximation
$Γ = Special::gammaStirling($z);  // Stirling approximation
$l = Special::logGamma($z);
$c = Special::logGammaCorr($z);   // Log gamma correction

// Incomplete gamma functions - γ(s,t), Γ(s,x), P(s,x)
[$x, $s] = [1, 2];
$γ = Special::lowerIncompleteGamma($x, $s);
$Γ = Special::upperIncompleteGamma($x, $s);
$P = Special::regularizedLowerIncompleteGamma($x, $s);

// Beta function
[$x, $y] = [1, 2];
$β  = Special::beta($x, $y);
$lβ = Special::logBeta($x, $y);

// Incomplete beta functions
[$x, $a, $b] = [0.4, 2, 3];
$B  = Special::incompleteBeta($x, $a, $b);
$Iₓ = Special::regularizedIncompleteBeta($x, $a, $b);

// Multivariate beta function
$αs = [1, 2, 3];
$β  = Special::multivariateBeta($αs);

// Error function (Gauss error function)
$error = Special::errorFunction(2);              // same as erf
$error = Special::erf(2);                        // same as errorFunction
$error = Special::complementaryErrorFunction(2); // same as erfc

CHANGELOG.md
# MathPHP Change Log

## v2.8.1 - 2023-05-18

### Improvements

* Internal improvements to improve conformance with static analysis tools

## v2.8.0 - 2023-05-07

### New Features

* Matrix `rowAddVector`
* Matrix `columnAddVector`

### Improvements

* Better error handling and exception message in `Sequence\NonIntenger::hyperharmonic`
* Internal code improvements to conform to static analysis checks

### Backwards Incompatible Changes

* Helper method names changed (public abstract methods but not part of published interface)
    * `NumericalDifferentiation::isTargetInPoints` changed to `assertTargetInPoints`
    * `NumericalDifferentiation::isSpacingConstant` changed to `assertSpacingConstant`

## v2.7.0 - 2022-12-31

### Improvements

* Improved algorithm for `regularizedIncompleteBeta`: Addresses issue 458
* Issue 456: Improved PHPDoc blocks: Changed "number" to "int|float"
* Added PHP 8.2 for CI test target

## v2.6.0 - 2022-04-10

### Improvements

* `Average::truncatedMean` behavior at 50% trim made consistent
* PHP 8.1 compatibility improvements

### Backwards Incompatible Changes

# `Average::truncatedMean` throws exception if trim percent greater than 50% rather than error or unpredictable results.

## v2.5.0 - 2021-11-21

### New Features

* Special function `logbeta`
* Special function `logGamma`
* Special function `logGammaCorr`
* Special function `stirlingError`

### Improvements

* Improvements in StudentT continuous distribution
* Improvements in special function `gamma`
* Improvements in special function `beta`

### Bug Fixes

* Issue 393 (regularizedIncompleteBeta NAN)
* Issue 429 (Linear regression CI division by zero)

## v2.4.0 - 2021-07-27

### New Features

* Complex Exponential (`exp`)
* Complex Exponentiation (`pow`)
* Zipf's Law Discrete Distribution
* Generalized harmonic non-integer sequence

### Improvements

* Fixed Complex `polarForm` to compute the right values
* Fixed `hyperharnomic` non-integer sequence. Previously was computing the wrong thing
* Fixed how `ArbitraryInterger` handles `pow` of negative exponents

### Backwards Incompatible Changes

* Complex `polarForm` now returns an array rather than a Complex number, as the Complex return was incorrect
* Interface to `hyperharmonic` non-integer sequence changed due to previous implementation being incorrect

## v2.3.0 - 2021-07-14

### New Features

* Matrix SVD (Singular Value Decomposition)
* Polynomial companion matrix

## v2.2.0 - 2021-07-11

### New Features

* PLS (Partial Least Squares Regression)

### Improvements

* Add custom `__debugInfo` to `NumericMatrix`

## v2.1.0 - 2021-07-07

### New Features

* Quaternion numbers

## v2.0.0 - 2021-05-09

### New Features

* Matrix Improvements
    * `walk` method to map a function to all values without mutation or returning a value
    * `MatrixFactory` creates more matrix types
    * `MatrixFactory::createNumeric` to create `NumericMatrix` types
    * `MatrixFactory::createFromRowVector`
    * `MatrixFactory::createFromColumnVector`
    * Internal `ObjectMatrix` improvements
        * Add `trace`
        * Add `scalarMultiply`
    * Add initial `ComplexMatrix`
* Sample data People

### Improvements

* Bug fixes
    * Issue 414 fixed - PCA/Eigenvalue convergence
    * Issue 413 fixed - matrix solve with singular matrix using RREF

## Migration - Upgrading to v2.0 from v1.0

* PHP minimum version now 7.2 (was 7.0)
* Deprecated code removed (backwards-incompatible change)
    * `MathPHP\Statistics\Distance::kullbackLeiblerDivergence` removed (
      Use `MathPHP\Statistics\Divergence::kullbackLeibler` instead)
    * `MathPHP\Statistics\Distance::jensenShannonDivergence` removed (Use `MathPHP\Statistics\Divergence::jensenShannon`
      instead)
    * Matrix Decompositions no longer implement `\ArrayAccess` interface to access decomposition matrixes. Use
      properties instead.
        * `MathPHP\LinearAlgebra\Decomposition\Cholesky`
            * `$cholesky['L']`, `$cholesky['Lᵀ']`, `$cholesky['LT']` removed,
              use `$cholesky->L`, `$cholesky->Lᵀ`, `$cholesky->LT` instead.
        * `MathPHP\LinearAlgebra\Decomposition\Crout`
            * `$crout['L']`, `$crout['U']` removed, use `$crout->L`, `$crout->U` instead.
        * `MathPHP\LinearAlgebra\Decomposition\LU`
            * `$LU['L']`, `LU['U']`, `LU['P']` removed, use `$LU->L`, `$LU->U`, `$LU->P` instead.
        * `MathPHP\LinearAlgebra\Decomposition\QR`
            * `$QR['Q']`, `$QR['R']` removed, use `$QR->Q`, `$QR->R` instead.
* Methods renamed (backwards-incompatible change)
    * `MathPHP\Statistics\Distance::bhattacharyyaDistance` renamed to `MathPHP\Statistics\Distance::bhattacharyya`
    * `MathPHP\Statistics\Distance::hellingerDistance` renamed to `MathPHP\Statistics\Distance::hellinger`
* Moved Functionality (backwards-incompatible change)
    * `MathPHP\Functions\Polynomial` moved to `MathPHP\Expression\Polynomial`
    * `MathPHP\Functions\Piecewise` moved to `MathPHP\Expression\Piecewise`
* Matrix internal refactoring
    * Note: These changes will not affect any client code as long as matrices were created using `MatrixFactory`.
    * `Matrix` is not a base abstract class for all matrix classes to extend
    * `Matrix` renamed `NumericMatrix`
    * `Matrix` base method `createZeroValue`
        * Use case is various `ObjectMatrix` classes that implement `ObjectArithmetic`
    * `RowVector` removed. Use `MatrixFactory::createFromRowVector` instead
    * `ColumnVector` removed. Use `MatrixFactory::createFromColumnVector` instead

## v1.11.0 - 2021-05-09

### Improvements

* Bugfix (Issue 413): Matrix solve with singular matrix using RREF
* Bugfix (Issue 414): PCA/Eigenvalue convergence

## v1.10.0 - 2020-12-19

### Improvements

* Bugfix (Issue 356): Fix Finance IRR NANs

## v1.9.0 - 2020-12-13

### New Features

* Vector min and max
* Arithmetic isqrt (integer square root)

### Improvements

* Remove Travis CI (Moved CI to Github Actions in v1.8.0 release)

src/SampleData/Iris.php
<?php

    namespace MathPHP\SampleData;

    use function array_column;
    use function array_values;

    /**
     * iris dataset (Edgar Anderson's Iris Data)
     *
     * This famous (Fisher's or Anderson's) iris data set gives the measurements in centimeters of the variables sepal length
     *  and width and petal length and width, respectively, for 50 flowers from each of 3 species of iris.
     * The species are Iris setosa, versicolor, and virginica.
     *
     * 150 observations on 5 variables: sepal length, sepal width, petal length, petal width, species.
     * R iris
     *
     * Source: Fisher, R. A. (1936) The use of multiple measurements in taxonomic problems. Annals of Eugenics, 7, Part II, 179–188.
     * The data were collected by Anderson, Edgar (1935). The irises of the Gaspe Peninsula, Bulletin of the American Iris Society, 59, 2–5.
     */
    class Iris {

        private const DATA
            = [
                [5.1, 3.5, 1.4, 0.2, 'setosa'],
                [4.9, 3.0, 1.4, 0.2, 'setosa'],
                [4.7, 3.2, 1.3, 0.2, 'setosa'],
                [4.6, 3.1, 1.5, 0.2, 'setosa'],
                [5.0, 3.6, 1.4, 0.2, 'setosa'],
                [5.4, 3.9, 1.7, 0.4, 'setosa'],
                [4.6, 3.4, 1.4, 0.3, 'setosa'],
                [5.0, 3.4, 1.5, 0.2, 'setosa'],
                [4.4, 2.9, 1.4, 0.2, 'setosa'],
                [4.9, 3.1, 1.5, 0.1, 'setosa'],
                [5.4, 3.7, 1.5, 0.2, 'setosa'],
                [4.8, 3.4, 1.6, 0.2, 'setosa'],
                [4.8, 3.0, 1.4, 0.1, 'setosa'],
                [4.3, 3.0, 1.1, 0.1, 'setosa'],
                [5.8, 4.0, 1.2, 0.2, 'setosa'],
                [5.7, 4.4, 1.5, 0.4, 'setosa'],
                [5.4, 3.9, 1.3, 0.4, 'setosa'],
                [5.1, 3.5, 1.4, 0.3, 'setosa'],
                [5.7, 3.8, 1.7, 0.3, 'setosa'],
                [5.1, 3.8, 1.5, 0.3, 'setosa'],
                [5.4, 3.4, 1.7, 0.2, 'setosa'],
                [5.1, 3.7, 1.5, 0.4, 'setosa'],
                [4.6, 3.6, 1.0, 0.2, 'setosa'],
                [5.1, 3.3, 1.7, 0.5, 'setosa'],
                [4.8, 3.4, 1.9, 0.2, 'setosa'],
                [5.0, 3.0, 1.6, 0.2, 'setosa'],
                [5.0, 3.4, 1.6, 0.4, 'setosa'],
                [5.2, 3.5, 1.5, 0.2, 'setosa'],
                [5.2, 3.4, 1.4, 0.2, 'setosa'],
                [4.7, 3.2, 1.6, 0.2, 'setosa'],
                [4.8, 3.1, 1.6, 0.2, 'setosa'],
                [5.4, 3.4, 1.5, 0.4, 'setosa'],
                [5.2, 4.1, 1.5, 0.1, 'setosa'],
                [5.5, 4.2, 1.4, 0.2, 'setosa'],
                [4.9, 3.1, 1.5, 0.2, 'setosa'],
                [5.0, 3.2, 1.2, 0.2, 'setosa'],
                [5.5, 3.5, 1.3, 0.2, 'setosa'],
                [4.9, 3.6, 1.4, 0.1, 'setosa'],
                [4.4, 3.0, 1.3, 0.2, 'setosa'],
                [5.1, 3.4, 1.5, 0.2, 'setosa'],
                [5.0, 3.5, 1.3, 0.3, 'setosa'],
                [4.5, 2.3, 1.3, 0.3, 'setosa'],
                [4.4, 3.2, 1.3, 0.2, 'setosa'],
                [5.0, 3.5, 1.6, 0.6, 'setosa'],
                [5.1, 3.8, 1.9, 0.4, 'setosa'],
                [4.8, 3.0, 1.4, 0.3, 'setosa'],
                [5.1, 3.8, 1.6, 0.2, 'setosa'],
                [4.6, 3.2, 1.4, 0.2, 'setosa'],
                [5.3, 3.7, 1.5, 0.2, 'setosa'],
                [5.0, 3.3, 1.4, 0.2, 'setosa'],
                [7.0, 3.2, 4.7, 1.4, 'versicolor'],
                [6.4, 3.2, 4.5, 1.5, 'versicolor'],
                [6.9, 3.1, 4.9, 1.5, 'versicolor'],
                [5.5, 2.3, 4.0, 1.3, 'versicolor'],
                [6.5, 2.8, 4.6, 1.5, 'versicolor'],
                [5.7, 2.8, 4.5, 1.3, 'versicolor'],
                [6.3, 3.3, 4.7, 1.6, 'versicolor'],
                [4.9, 2.4, 3.3, 1.0, 'versicolor'],
                [6.6, 2.9, 4.6, 1.3, 'versicolor'],
                [5.2, 2.7, 3.9, 1.4, 'versicolor'],
                [5.0, 2.0, 3.5, 1.0, 'versicolor'],
                [5.9, 3.0, 4.2, 1.5, 'versicolor'],
                [6.0, 2.2, 4.0, 1.0, 'versicolor'],
                [6.1, 2.9, 4.7, 1.4, 'versicolor'],
                [5.6, 2.9, 3.6, 1.3, 'versicolor'],
                [6.7, 3.1, 4.4, 1.4, 'versicolor'],
                [5.6, 3.0, 4.5, 1.5, 'versicolor'],
                [5.8, 2.7, 4.1, 1.0, 'versicolor'],
                [6.2, 2.2, 4.5, 1.5, 'versicolor'],
                [5.6, 2.5, 3.9, 1.1, 'versicolor'],
                [5.9, 3.2, 4.8, 1.8, 'versicolor'],
                [6.1, 2.8, 4.0, 1.3, 'versicolor'],
                [6.3, 2.5, 4.9, 1.5, 'versicolor'],
                [6.1, 2.8, 4.7, 1.2, 'versicolor'],
                [6.4, 2.9, 4.3, 1.3, 'versicolor'],
                [6.6, 3.0, 4.4, 1.4, 'versicolor'],
                [6.8, 2.8, 4.8, 1.4, 'versicolor'],
                [6.7, 3.0, 5.0, 1.7, 'versicolor'],
                [6.0, 2.9, 4.5, 1.5, 'versicolor'],
                [5.7, 2.6, 3.5, 1.0, 'versicolor'],
                [5.5, 2.4, 3.8, 1.1, 'versicolor'],
                [5.5, 2.4, 3.7, 1.0, 'versicolor'],
                [5.8, 2.7, 3.9, 1.2, 'versicolor'],
                [6.0, 2.7, 5.1, 1.6, 'versicolor'],
                [5.4, 3.0, 4.5, 1.5, 'versicolor'],
                [6.0, 3.4, 4.5, 1.6, 'versicolor'],
                [6.7, 3.1, 4.7, 1.5, 'versicolor'],
                [6.3, 2.3, 4.4, 1.3, 'versicolor'],
                [5.6, 3.0, 4.1, 1.3, 'versicolor'],
                [5.5, 2.5, 4.0, 1.3, 'versicolor'],
                [5.5, 2.6, 4.4, 1.2, 'versicolor'],
                [6.1, 3.0, 4.6, 1.4, 'versicolor'],
                [5.8, 2.6, 4.0, 1.2, 'versicolor'],
                [5.0, 2.3, 3.3, 1.0, 'versicolor'],
                [5.6, 2.7, 4.2, 1.3, 'versicolor'],
                [5.7, 3.0, 4.2, 1.2, 'versicolor'],
                [5.7, 2.9, 4.2, 1.3, 'versicolor'],
                [6.2, 2.9, 4.3, 1.3, 'versicolor'],
                [5.1, 2.5, 3.0, 1.1, 'versicolor'],
                [5.7, 2.8, 4.1, 1.3, 'versicolor'],
                [6.3, 3.3, 6.0, 2.5, 'virginica'],
                [5.8, 2.7, 5.1, 1.9, 'virginica'],
                [7.1, 3.0, 5.9, 2.1, 'virginica'],
                [6.3, 2.9, 5.6, 1.8, 'virginica'],
                [6.5, 3.0, 5.8, 2.2, 'virginica'],
                [7.6, 3.0, 6.6, 2.1, 'virginica'],
                [4.9, 2.5, 4.5, 1.7, 'virginica'],
                [7.3, 2.9, 6.3, 1.8, 'virginica'],
                [6.7, 2.5, 5.8, 1.8, 'virginica'],
                [7.2, 3.6, 6.1, 2.5, 'virginica'],
                [6.5, 3.2, 5.1, 2.0, 'virginica'],
                [6.4, 2.7, 5.3, 1.9, 'virginica'],
                [6.8, 3.0, 5.5, 2.1, 'virginica'],
                [5.7, 2.5, 5.0, 2.0, 'virginica'],
                [5.8, 2.8, 5.1, 2.4, 'virginica'],
                [6.4, 3.2, 5.3, 2.3, 'virginica'],
                [6.5, 3.0, 5.5, 1.8, 'virginica'],
                [7.7, 3.8, 6.7, 2.2, 'virginica'],
                [7.7, 2.6, 6.9, 2.3, 'virginica'],
                [6.0, 2.2, 5.0, 1.5, 'virginica'],
                [6.9, 3.2, 5.7, 2.3, 'virginica'],
                [5.6, 2.8, 4.9, 2.0, 'virginica'],
                [7.7, 2.8, 6.7, 2.0, 'virginica'],
                [6.3, 2.7, 4.9, 1.8, 'virginica'],
                [6.7, 3.3, 5.7, 2.1, 'virginica'],
                [7.2, 3.2, 6.0, 1.8, 'virginica'],
                [6.2, 2.8, 4.8, 1.8, 'virginica'],
                [6.1, 3.0, 4.9, 1.8, 'virginica'],
                [6.4, 2.8, 5.6, 2.1, 'virginica'],
                [7.2, 3.0, 5.8, 1.6, 'virginica'],
                [7.4, 2.8, 6.1, 1.9, 'virginica'],
                [7.9, 3.8, 6.4, 2.0, 'virginica'],
                [6.4, 2.8, 5.6, 2.2, 'virginica'],
                [6.3, 2.8, 5.1, 1.5, 'virginica'],
                [6.1, 2.6, 5.6, 1.4, 'virginica'],
                [7.7, 3.0, 6.1, 2.3, 'virginica'],
                [6.3, 3.4, 5.6, 2.4, 'virginica'],
                [6.4, 3.1, 5.5, 1.8, 'virginica'],
                [6.0, 3.0, 4.8, 1.8, 'virginica'],
                [6.9, 3.1, 5.4, 2.1, 'virginica'],
                [6.7, 3.1, 5.6, 2.4, 'virginica'],
                [6.9, 3.1, 5.1, 2.3, 'virginica'],
                [5.8, 2.7, 5.1, 1.9, 'virginica'],
                [6.8, 3.2, 5.9, 2.3, 'virginica'],
                [6.7, 3.3, 5.7, 2.5, 'virginica'],
                [6.7, 3.0, 5.2, 2.3, 'virginica'],
                [6.3, 2.5, 5.0, 1.9, 'virginica'],
                [6.5, 3.0, 5.2, 2.0, 'virginica'],
                [6.2, 3.4, 5.4, 2.3, 'virginica'],
                [5.9, 3.0, 5.1, 1.8, 'virginica'],
            ];

        /**
         * Raw data without labels
         * [[5.1, 3.5, 1.4, 0.2, 'setosa'], [4.9, 3.0, 1.4, 0.2, 'setosa'], ... ]
         *
         * @return array[]
         */
        public static function getData(): array
        {
            return array_values(self::DATA);
        }

        /**
         * Raw data with each observation labeled
         * [['sepalLength' => 5.11, 'sepalWidth' => 3.5, 'petalLength' => 1.4, 'petalWidth' => 0.2, 'species' => 'setosa'], ... ]
         *
         * @return array<array<string, float>>
         */
        public static function getLabeledData(): array
        {
            /** @var array<array<string, float>> $array_map */
            $array_map = [];
            foreach (self::DATA as $ignored => array $data)

            return $array_map;
        }

        /**
         * Sepal length observations
         *
         * @return float[]
         */
        public static function getSepalLength(): array
        {
            return array_column(self::DATA, 0);
        }

        /**
         * Sepal width observations
         *
         * @return float[]
         */
        public static function getSepalWidth(): array
        {
            return array_column(self::DATA, 1);
        }

        /**
         * Petal length observations
         *
         * @return float[]
         */
        public static function getPetalLength(): array
        {
            return array_column(self::DATA, 2);
        }

        /**
         * Petal width observations
         *
         * @return float[]
         */
        public static function getPetalWidth(): array
        {
            return array_column(self::DATA, 3);
        }

        /**
         * Species observations
         *
         * @return string[]
         */
        public static function getSpecies(): array
        {
            return array_column(self::DATA, 4);
        }

        public static function numberOfSpecies()
        {
        }

        public static function numberOfPetalWidth()
        {
        }

        public static function numberOfPetalLength()
        {
        }

        public static function numberOfSepalWidth()
        {
        }

        public static function numberOfSepalLength()
        {
        }

        public static function labeledData()
        {
        }

        public static function dataHas5Variables()
        {
        }

        public static function dataHas150Observations()
        {
        }
    }
